#  Coursera Capstone Project : Segmenting and Clustering Neighborhoods in Toronto

 **This Notebook contains the Coursera Data Science Capstone Project - by Krishanu Podder** 

#### Importing Neccessary Libraries

In [1]:
import numpy as np

import pandas as pd
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json

!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim 

import requests 
from pandas.io.json import json_normalize 

import matplotlib.cm as cm
import matplotlib.colors as colors

from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes 
import folium 

print('Libraries imported.')

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    python_abi-3.6             |          1_cp36m           4 KB  conda-forge
    geographiclib-1.50         |             py_0          34 KB  conda-forge
    geopy-1.22.0               |     pyh9f0ad1d_0          63 KB  conda-forge
    ca-certificates-2020.4.5.1 |       hecc5488_0         146 KB  conda-forge
    openssl-1.1.1g             |       h516909a_0         2.1 MB  conda-forge
    certifi-2020.4.5.1         |   py36h9f0ad1d_0         151 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         2.5 MB

The following NEW packages will be INSTALLED:

    geographiclib:   1.50-py_0           conda-forge
    geopy:          

# Question 1:

# Use pandas, or the BeautifulSoup package, or any other way you are comfortable with to transform the data in the table on the Wikipedia page into the above pandas dataframe.

#### Getting data from wikipedia page

In [11]:
link = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'

df_old = pd.read_html(link)[0]
df_old

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
7           M8A      Not assigned   
8           M9A         Etobicoke   
9           M1B       Scarborough   
10          M2B      Not assigned   
11          M3B        North York   
12          M4B         East York   
13          M5B  Downtown Toronto   
14          M6B        North York   
15          M7B      Not assigned   
16          M8B      Not assigned   
17          M9B         Etobicoke   
18          M1C       Scarborough   
19          M2C      Not assigned   
20          M3C        North York   
21          M4C         East York   
22          M5C  Downtown Toronto   
23          M6C              York   
24          M7C      Not assigned   
25          M8C      Not assigned   
26          M9C         Etobicoke   
27          M1E       Scarborough   
28          M2E      Not assigned   
29          M3E      Not assigned   
30          M4E      East Toronto   
31          M5E  Downtown Toronto   
32          M6E              York   
33          M7E      Not assigned   
34          M8E      Not assigned   
35          M9E      Not assigned   
36          M1G       Scarborough   
37          M2G      Not assigned   
38          M3G      Not assigned   
39          M4G         East York   
40          M5G  Downtown Toronto   
41          M6G  Downtown Toronto   
42          M7G      Not assigned   
43          M8G      Not assigned   
44          M9G      Not assigned   
45          M1H       Scarborough   
46          M2H        North York   
47          M3H        North York   
48          M4H         East York   
49          M5H  Downtown Toronto   
50          M6H      West Toronto   
51          M7H      Not assigned   
52          M8H      Not assigned   
53          M9H      Not assigned   
54          M1J       Scarborough   
55          M2J        North York   
56          M3J        North York   
57          M4J         East York   
58          M5J  Downtown Toronto   
59          M6J      West Toronto   
60          M7J      Not assigned   
61          M8J      Not assigned   
62          M9J      Not assigned   
63          M1K       Scarborough   
64          M2K        North York   
65          M3K        North York   
66          M4K      East Toronto   
67          M5K  Downtown Toronto   
68          M6K      West Toronto   
69          M7K      Not assigned   
70          M8K      Not assigned   
71          M9K      Not assigned   
72          M1L       Scarborough   
73          M2L        North York   
74          M3L        North York   
75          M4L      East Toronto   
76          M5L  Downtown Toronto   
77          M6L        North York   
78          M7L      Not assigned   
79          M8L      Not assigned   
80          M9L        North York   
81          M1M       Scarborough   
82          M2M        North York   
83          M3M        North York   
84          M4M      East Toronto   
85          M5M        North York   
86          M6M              York   
87          M7M      Not assigned   
88          M8M      Not assigned   
89          M9M        North York   
90          M1N       Scarborough   
91          M2N        North York   
92          M3N        North York   
93          M4N   Central Toronto   
94          M5N   Central Toronto   
95          M6N              York   
96          M7N      Not assigned   
97          M8N      Not assigned   
98          M9N              York   
99          M1P       Scarborough   
100         M2P        North York   
101         M3P      Not assigned   
102         M4P   Central Toronto   
103         M5P   Central Toronto   
104         M6P      West Toronto   
105         M7P      Not assigned   
106         M8P      Not assigned   
107     

#### Finding dimension of uncleaned data

In [3]:
df_old.shape

(180, 3)

#### Ignoring cells with a borough that is Not assigned.

In [4]:
df1= df_old[df_old.Borough!='Not assigned']
df1.reset_index(drop=True)
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(df1['Borough'].unique()),
        df1.shape[0]
    )
)

The dataframe has 10 boroughs and 103 neighborhoods.


#### Finding shape of the new dataframe

In [5]:
df1.shape

(103, 3)

## So, the new dataframe has 103 rows and 3 columns

# Question 2:

# Use the Geocoder package or the csv file to create dataframe with longitude and latitude values

#### reading csv file that has the geographical coordinates of each postal code

In [7]:
df2 = pd.read_csv('https://cocl.us/Geospatial_data')

#### Merging the dataframe 'df1' (having Postal Code, Borough, Neighborhood) with dataframe 'df2' (having coordinates)

In [8]:
df = df1.merge(df2, on='Postal Code')

#### Glimpse of the new dataframe 'df'

In [9]:
df.head()

Postal Code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

#### Getting the shape of the new dataframe 'df'

In [10]:
df.shape

(103, 5)

### Well, the new dataframe has 103 rows and 5 columns. So the two dataframe has been merged successfully.

# Question 3:

# Explore and cluster the neighborhoods in Toronto

In [12]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(df['Borough'].unique()),
        df.shape[0]
    )
)

The dataframe has 10 boroughs and 103 neighborhoods.


#### Well, we have seen the dataframe has 10 boroughs and 103 neighborhoods

#### We use geopy library to get the latitude and longitude values of Toronto City.
#### In order to define an instance of the geocoder, we need to define a user_agent. We will name our agent ny_explorer, as shown below.

In [13]:
address = 'Toronto'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


#### We will Create a map of Toronto with neighborhoods superimposed on top.

In [14]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df['Latitude'], df['Longitude'], df['Borough'], df['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

### Incase the map_toronto is not visible, I have also uploaded an image file (d1) in github. Please refer that.

#### for illustration purposes, let's simplify the above map and segment and cluster only the neighborhoods in Downtown Toronto. So let's slice the original dataframe and create a new dataframe of the Downtown Toronto.

In [15]:
downtown_toronto = df[df.Borough=='Downtown Toronto'].reset_index(drop=True)
downtown_toronto.head()

Postal Code           Borough                                 Neighborhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M5E  Downtown Toronto                                  Berczy Park   

    Latitude  Longitude  
0  43.654260 -79.360636  
1  43.662301 -79.389494  
2  43.657162 -79.378937  
3  43.651494 -79.375418  
4  43.644771 -79.373306

#### Let's get the geographical coordinates of Downtown Toronto.

In [16]:
address = 'Downtown Toronto'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Downtown Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Downtown Toronto are 43.6541737, -79.38081164513409.


#### let's visualize Downtown Toronto (dt), the neighborhoods in it.

In [17]:
# create map of Downtown Toronto using latitude and longitude values
map_dt = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(downtown_toronto['Latitude'], downtown_toronto['Longitude'], downtown_toronto['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_dt)  
    
map_dt

#### Incase the map_dt is not visible, I have also uploaded an image file (d2) in github. Please refer that.

#### Next, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them.

#### Defining Foursquare Credentials and Version (Cell is hidden as it contains my Client ID & Secret)

In [45]:
# The code was removed by Watson Studio for sharing.

Your credentails:
CLIENT_ID: QWA2CYDEKZLVI1PVHE1IDR0EJKZ5UWHKMOT50WB3NKYZQOOM
CLIENT_SECRET:I01NYJHPR1ZMPVAYDWAADN2ULD3TKDBWH04UES320EOBJGBE


#### Let's explore the first neighborhood in our dataframe.
#### Get the neighborhood's name.

In [19]:
downtown_toronto.loc[0, 'Neighborhood']

'Regent Park, Harbourfront'

#### Getting the neighborhood's latitude and longitude values.

In [20]:
neighborhood_latitude = downtown_toronto.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = downtown_toronto.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = downtown_toronto.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Regent Park, Harbourfront are 43.6542599, -79.3606359.


#### Now, let's get the top 100 venues that are in Regent Park, Harbourfront within a radius of 500 meters.
#### First, let's create the GET request URL. Name your URL url.

In [21]:
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url

'https://api.foursquare.com/v2/venues/explore?&client_id=QWA2CYDEKZLVI1PVHE1IDR0EJKZ5UWHKMOT50WB3NKYZQOOM&client_secret=I01NYJHPR1ZMPVAYDWAADN2ULD3TKDBWH04UES320EOBJGBE&v=20180605&ll=43.6542599,-79.3606359&radius=500&limit=100'

#### Sending the GET request and examining the results

In [22]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5eca5854fb34b5001b12f51b'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Corktown',
  'headerFullLocation': 'Corktown, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 48,
  'suggestedBounds': {'ne': {'lat': 43.6587599045, 'lng': -79.3544279001486},
   'sw': {'lat': 43.6497598955, 'lng': -79.36684389985142}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '54ea41ad498e9a11e9e13308',
       'name': 'Roselle Desserts',
       'location': {'address': '362 King St E',
        'crossStreet': 'Trinity St',
        'lat': 43.653446723052674,
        'lng': -79.3620167174383,
        'labeledLatLngs': [{'label': 'display',
 

#### From the Foursquare lab , we know that all the information is in the items key. Before we proceed, let's borrow the get_category_type function from the Foursquare lab.

In [23]:
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

#### Now we are ready to clean the json and structure it into a pandas dataframe.

In [24]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name           categories        lat        lng
0        Roselle Desserts               Bakery  43.653447 -79.362017
1           Tandem Coffee          Coffee Shop  43.653559 -79.361809
2      Morning Glory Cafe       Breakfast Spot  43.653947 -79.361149
3  Cooper Koo Family YMCA  Distribution Center  43.653249 -79.358008
4     Body Blitz Spa East                  Spa  43.654735 -79.359874

In [25]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

48 venues were returned by Foursquare.


#### Exploring Neighborhoods in Downtown Toronto. Let's create a function to repeat the same process to all the neighborhoods in Downtown toronto

In [26]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    LIMIT = 100
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

####  creating a new dataframe called dt_venues

In [27]:
dt_venues = getNearbyVenues(names=downtown_toronto['Neighborhood'],
                                   latitudes=downtown_toronto['Latitude'],
                                   longitudes=downtown_toronto['Longitude']
                                  )

Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Harbourfront East, Union Station, Toronto Islands
Toronto Dominion Centre, Design Exchange
Commerce Court, Victoria Hotel
University of Toronto, Harbord
Kensington Market, Chinatown, Grange Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Rosedale
Stn A PO Boxes
St. James Town, Cabbagetown
First Canadian Place, Underground city
Church and Wellesley


#### Let's check the size of the resulting dataframe

In [28]:
print(dt_venues.shape)
dt_venues.head()

(1211, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park, Harbourfront               43.65426              -79.360636   
1  Regent Park, Harbourfront               43.65426              -79.360636   
2  Regent Park, Harbourfront               43.65426              -79.360636   
3  Regent Park, Harbourfront               43.65426              -79.360636   
4  Regent Park, Harbourfront               43.65426              -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2      Morning Glory Cafe       43.653947       -79.361149   
3  Cooper Koo Family YMCA       43.653249       -79.358008   
4     Body Blitz Spa East       43.654735       -79.359874   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2       Breakfast Spot  
3  Distribution Center  
4                  Spa

#### Let's check how many venues were returned for each neighborhood

In [29]:
dt_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            56   
CN Tower, King and Spadina, Railway Lands, Harb...                     15   
Central Bay Street                                                     62   
Christie                                                               17   
Church and Wellesley                                                   76   
Commerce Court, Victoria Hotel                                        100   
First Canadian Place, Underground city                                100   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
Kensington Market, Chinatown, Grange Park                              58   
Queen's Park, Ontario Provincial Government                            34   
Regent Park, Harbourfront                                              48   
Richmond, Adelaide, King                                               92   
Rosedale                                                                4   
St. James Town                                                         77   
St. James Town, Cabbagetown                                            46   
Stn A PO Boxes                                                         92   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         34   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             56   
CN Tower, King and Spadina, Railway Lands, Harb...                      15   
Central Bay Street                                                      62   
Christie                                                                17   
Church and Wellesley                                                    76   
Commerce Court, Victoria Hotel                                         100   
First Canadian Place, Underground city                                 100   
Garden District, Ryerson                                               100   
Harbourfront East, Union Station, Toronto Islands                      100   
Kensington Market, Chinatown, Grange Park                               58   
Queen's Park, Ontario Provincial Government                             34   
Regent Park, Harbourfront                                               48   
Richmond, Adelaide, King                                                92   
Rosedale                                                                 4   
St. James Town                                                          77   
St. James Town, Cabbagetown                                             46   
Stn A PO Boxes                                                          92   
Toronto Dominion Centre, Design Exchange                               100   
University of Toronto, Harbord                                          34   

                                                    Venue  Venue Latitude  \
Neighborhood                                                                
Berczy Park                                            56              56   
CN Tower, King and Spadina, Railway Lands, Harb...     15              15   
Central Bay Street                                     62              62   
Christie                                               17              17   
Church and Wellesley                                   76              76   
Commerce Court, Victoria Hotel                        100             100   
First Canadian Place, Underground city                100             100   
Garden District, Ryerson                              100             100   
Harbourfront East, Union 

#### Let's find out how many unique categories can be curated from all the returned venues

In [30]:
print('There are {} uniques categories.'.format(len(dt_venues['Venue Category'].unique())))

There are 209 uniques categories.


#### Analyzing Each Neighborhood

In [31]:
# one hot encoding
toronto_onehot = pd.get_dummies(dt_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = dt_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0            0                  0        0                   0             0   
1            0                  0        0                   0             0   
2            0                  0        0                   0             0   
3            0                  0        0                   0             0   
4            0                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0             0         0            0           0                    0   
1             0         0            0           0                    0   
2             0         0            0           0                    0   
3             0         0            0           0                    0   
4             0         0            0           0                    0   

   Asian Restaurant  Athletics & Sports  BBQ Joint  Baby Store  Bagel Shop  \
0                 0                   0          0           0           0   
1                 0                   0          0           0           0   
2                 0                   0          0           0           0   
3                 0                   0          0           0           0   
4                 0                   0          0           0           0   

   Bakery  Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  \
0       1     0    0                 0                   0      0   
1       0     0    0                 0                   0      0   
2       0     0    0                 0                   0      0   
3       0     0    0                 0                   0      0   
4       0     0    0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bistro  \
0                0         0           0                   0       0   
1                0         0           0                   0       0   
2                0         0           0                   0       0   
3                0         0           0                   0       0   
4                0         0           0                   0       0   

   Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  \
0              0          0         0                     0               0   
1              0          0         0                     0               0   
2              0          0         0                     0               1   
3              0          0         0                     0               0   
4              0          0         0                     0               0   

   Brewery  Bubble Tea Shop  Building  Burger Joint  Burrito Place  Butcher  \
0        0                0         0             0              0        0   
1        0                0         0             0              0        0   
2        0                0         0             0              0        0   
3        0                0         0             0              0        0   
4        0                0         0             0              0        0   

   Café  Camera Store  Candy Store  Caribbean Restaurant  Cheese Shop  \
0     0             0            0                     0            0   
1     0             0            0                     0            0   
2     0             0            0                     0            0   
3     0             0            0               

#### And let's examine the new dataframe size.

In [32]:
toronto_onehot.shape

(1211, 209)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [33]:
dt_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
dt_grouped

Neighborhood  Yoga Studio  \
0                                         Berczy Park     0.000000   
1   CN Tower, King and Spadina, Railway Lands, Har...     0.000000   
2                                  Central Bay Street     0.016129   
3                                            Christie     0.000000   
4                                Church and Wellesley     0.026316   
5                      Commerce Court, Victoria Hotel     0.000000   
6              First Canadian Place, Underground city     0.000000   
7                            Garden District, Ryerson     0.000000   
8   Harbourfront East, Union Station, Toronto Islands     0.000000   
9           Kensington Market, Chinatown, Grange Park     0.000000   
10        Queen's Park, Ontario Provincial Government     0.029412   
11                          Regent Park, Harbourfront     0.020833   
12                           Richmond, Adelaide, King     0.000000   
13                                           Rosedale     0.000000   
14                                     St. James Town     0.000000   
15                        St. James Town, Cabbagetown     0.000000   
16                                     Stn A PO Boxes     0.010870   
17           Toronto Dominion Centre, Design Exchange     0.000000   
18                     University of Toronto, Harbord     0.029412   

    Afghan Restaurant   Airport  Airport Food Court  Airport Gate  \
0            0.000000  0.000000            0.000000      0.000000   
1            0.000000  0.066667            0.066667      0.066667   
2            0.000000  0.000000            0.000000      0.000000   
3            0.000000  0.000000            0.000000      0.000000   
4            0.013158  0.000000            0.000000      0.000000   
5            0.000000  0.000000            0.000000      0.000000   
6            0.000000  0.000000            0.000000      0.000000   
7            0.000000  0.000000            0.000000      0.000000   
8            0.000000  0.000000            0.000000      0.000000   
9            0.000000  0.000000            0.000000      0.000000   
10           0.000000  0.000000            0.000000      0.000000   
11           0.000000  0.000000            0.000000      0.000000   
12           0.000000  0.000000            0.000000      0.000000   
13           0.000000  0.000000            0.000000      0.000000   
14           0.000000  0.000000            0.000000      0.000000   
15           0.000000  0.000000            0.000000      0.000000   
16           0.000000  0.000000            0.000000      0.000000   
17           0.000000  0.000000            0.000000      0.000000   
18           0.000000  0.000000            0.000000      0.000000   

    Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0         0.000000         0.000000          0.000000             0.000000   
1         0.133333         0.133333          0.133333             0.000000   
2         0.000000         0.000000          0.000000             0.000000   
3         0.000000         0.000000          0.000000             0.000000   
4         0.000000         0.000000          0.000000             0.013158   
5         0.000000         0.000000          0.000000             0.040000   
6         0.000000         0.000000          0.000000             0.030000   
7         0.000000         0.000000          0.000000             0.000000   
8         0.000000         0.000000          0.000000             0.000000   
9         0.000000         0.000000          0.000000             0.000000   
10        0.000000         0.000000          0.000000             0.000000   
11        0.000000         0.000000          0.000000             0.000000   
12        0.000000         0.000000          0.000000             0.021739   
13        0.000000         0.000000          0.000000             0.000000   
14        0.000000         0.000000          0.000000             0.038961   
15        0

#### Let's print each neighborhood along with the top 5 most common venues

In [34]:
num_top_venues = 5

for hood in dt_grouped['Neighborhood']:
#     print("----"+hood+"----")
    temp = dt_grouped[dt_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

                venue  freq
0         Coffee Shop  0.07
1        Cocktail Bar  0.05
2                Café  0.04
3  Seafood Restaurant  0.04
4         Cheese Shop  0.04


              venue  freq
0    Airport Lounge  0.13
1   Airport Service  0.13
2  Airport Terminal  0.13
3     Boat or Ferry  0.07
4           Airport  0.07


                venue  freq
0         Coffee Shop  0.18
1  Italian Restaurant  0.06
2      Sandwich Place  0.05
3                Café  0.05
4     Bubble Tea Shop  0.03


           venue  freq
0  Grocery Store  0.24
1           Café  0.18
2           Park  0.12
3     Baby Store  0.06
4      Nightclub  0.06


                 venue  freq
0          Coffee Shop  0.08
1  Japanese Restaurant  0.07
2     Sushi Restaurant  0.05
3           Restaurant  0.04
4              Gay Bar  0.04


                 venue  freq
0          Coffee Shop  0.11
1           Restaurant  0.07
2                 Café  0.07
3                Hotel  0.05
4  American Restaurant  0.04


          

#### Let's put that into a pandas dataframe
#### First, let's write a function to sort the venues in descending order.

In [35]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

#### Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [36]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = dt_grouped['Neighborhood']

for ind in np.arange(dt_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(dt_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1  CN Tower, King and Spadina, Railway Lands, Har...        Airport Lounge   
2                                 Central Bay Street           Coffee Shop   
3                                           Christie         Grocery Store   
4                               Church and Wellesley           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0          Cocktail Bar           Cheese Shop              Beer Bar   
1       Airport Service      Airport Terminal      Sculpture Garden   
2    Italian Restaurant        Sandwich Place                  Café   
3                  Café                  Park            Baby Store   
4   Japanese Restaurant      Sushi Restaurant            Restaurant   

  5th Most Common Venue 6th Most Common Venue     7th Most Common Venue  \
0            Restaurant                Bakery                      Café   
1               Airport    Airport Food Court              Airport Gate   
2       Thai Restaurant           Salad Place                       Bar   
3           Coffee Shop             Nightclub        Athletics & Sports   
4               Gay Bar                  Café  Mediterranean Restaurant   

  8th Most Common Venue        9th Most Common Venue 10th Most Common Venue  
0    Seafood Restaurant  Eastern European Restaurant          Grocery Store  
1         Boat or Ferry                          Bar        Harbor / Marina  
2       Bubble Tea Shop                 Burger Joint         Ice Cream Shop  
3                 Diner                   Restaurant     Italian Restaurant  
4           Men's Store                        Hotel            Yoga Studio

#### Cluster Neighborhoods. Running k-means to cluster the neighborhood into 5 clusters.

In [37]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = dt_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 3, 4, 2, 0, 0, 0, 0, 0, 0], dtype=int32)

#### Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [38]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = downtown_toronto

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

Postal Code           Borough                                 Neighborhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M5E  Downtown Toronto                                  Berczy Park   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.654260 -79.360636               0           Coffee Shop   
1  43.662301 -79.389494               4           Coffee Shop   
2  43.657162 -79.378937               0        Clothing Store   
3  43.651494 -79.375418               0           Coffee Shop   
4  43.644771 -79.373306               0           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                Bakery                  Park                   Pub   
1           Yoga Studio                 Diner                   Bar   
2           Coffee Shop       Bubble Tea Shop        Cosmetics Shop   
3                  Café          Cocktail Bar             Gastropub   
4          Cocktail Bar           Cheese Shop              Beer Bar   

  5th Most Common Venue      6th Most Common Venue 7th Most Common Venue  \
0        Breakfast Spot                       Café               Theater   
1         Smoothie Shop         Italian Restaurant   Japanese Restaurant   
2                  Café  Middle Eastern Restaurant    Italian Restaurant   
3   American Restaurant                        Gym      Department Store   
4            Restaurant                     Bakery                  Café   

  8th Most Common Venue        9th Most Common Venue 10th Most Common Venue  
0   Distribution Center                   Playground      Electronics Store  
1        Sandwich Place          Distribution Center          Burrito Place  
2            Restaurant          Japanese Restaurant                Theater  
3            Restaurant               Clothing Store               Beer Bar  
4    Seafood Restaurant  Eastern European Restaurant          Grocery Store

#### Finally, let's visualize the resulting clusters

In [39]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

#### Incase the map_clusters is not visible, I have also uploaded an image file (d3) in github. Please refer that

#### Examining Clusters


#### Now, we can examine each cluster and determine the discriminating venue categories that distinguish each cluster. Based on the defining categories, we can then assign a name to each cluster. 



#### Cluster 1

In [40]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0   Downtown Toronto               0           Coffee Shop   
2   Downtown Toronto               0        Clothing Store   
3   Downtown Toronto               0           Coffee Shop   
4   Downtown Toronto               0           Coffee Shop   
7   Downtown Toronto               0           Coffee Shop   
8   Downtown Toronto               0           Coffee Shop   
9   Downtown Toronto               0           Coffee Shop   
10  Downtown Toronto               0           Coffee Shop   
11  Downtown Toronto               0                  Café   
12  Downtown Toronto               0                  Café   
15  Downtown Toronto               0           Coffee Shop   
16  Downtown Toronto               0           Coffee Shop   
17  Downtown Toronto               0           Coffee Shop   
18  Downtown Toronto               0           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                 Bakery                  Park                   Pub   
2            Coffee Shop       Bubble Tea Shop        Cosmetics Shop   
3                   Café          Cocktail Bar             Gastropub   
4           Cocktail Bar           Cheese Shop              Beer Bar   
7                   Café            Restaurant                   Gym   
8               Aquarium                  Café                 Hotel   
9                  Hotel                  Café            Restaurant   
10                  Café            Restaurant                 Hotel   
11            Restaurant                   Bar    Italian Restaurant   
12          Dessert Shop           Coffee Shop    Mexican Restaurant   
15                  Café          Cocktail Bar   Japanese Restaurant   
16                  Café           Pizza Place                Bakery   
17                  Café            Restaurant                   Gym   
18   Japanese Restaurant      Sushi Restaurant            Restaurant   

    5th Most Common Venue      6th Most Common Venue  \
0          Breakfast Spot                       Café   
2                    Café  Middle Eastern Restaurant   
3     American Restaurant                        Gym   
4              Restaurant                     Bakery   
7          Clothing Store                      Hotel   
8              Restaurant         Italian Restaurant   
9             Salad Place        Japanese Restaurant   
10    American Restaurant                        Gym   
11    Japanese Restaurant                  Bookstore   
12  Vietnamese Restaurant                     Bakery   
15             Restaurant         Italian Restaurant   
16             Restaurant         Italian Restaurant   
17    Japanese Restaurant                      Hotel   
18                Gay Bar                       Café   

       7th Most Common Venue          8th Most Common Venue  \
0                    Theater            Distribution Center   
2         Italian Restaurant                     Restaurant   
3           Department Store                     Restaurant   
4                       Café             Seafood Restaurant   
7            Thai Restaurant                  Deli / Bodega   
8                    Brewery            Fried Chicken Joint   
9         Seafood Restaurant            American Restaurant   
10       Japanese Restaurant             Seafood Restaurant   
11                    Bakery                    Yoga Studio   
12                       Bar  Vegetarian / Vegan Restaurant   
15                  Beer Bar             Seafood Restaurant   
16                       Pub                     Playground   
17       American Restaurant               Asian Restaurant   
18  Mediterranean Restaurant                    Men's Store   

          9th Most Common Venue 10th Most Common Venue  
0                    Playground      Electronics Store  
2           Japanese Restaurant                Theater  
3                Clothing Store            

#### Cluster 2

In [41]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
14  Downtown Toronto               1                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
14            Playground                 Trail        Cosmetics Shop   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
14               Dog Run   Distribution Center        Discount Store   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
14                 Diner          Dessert Shop       Department Store

#### Cluster 3

In [42]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
6  Downtown Toronto               2         Grocery Store   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
6                  Café                  Park            Baby Store   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
6           Coffee Shop             Nightclub    Athletics & Sports   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
6                 Diner            Restaurant     Italian Restaurant

#### Cluster 4

In [43]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
13  Downtown Toronto               3        Airport Lounge   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
13       Airport Service      Airport Terminal      Sculpture Garden   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
13               Airport    Airport Food Court          Airport Gate   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
13         Boat or Ferry                   Bar        Harbor / Marina

#### Cluster 5

In [44]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
1  Downtown Toronto               4           Coffee Shop   
5  Downtown Toronto               4           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
1           Yoga Studio                 Diner                   Bar   
5    Italian Restaurant        Sandwich Place                  Café   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
1         Smoothie Shop    Italian Restaurant   Japanese Restaurant   
5       Thai Restaurant           Salad Place                   Bar   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
1        Sandwich Place   Distribution Center          Burrito Place  
5       Bubble Tea Shop          Burger Joint         Ice Cream Shop

## This project has been submitted by Krishanu Podder https://www.linkedin.com/in/krishanu-podder-905a9a89/